### DISCUSSION TITLE
I'm still really missing the AST-based navigation from Merlin / Vim, nearly every day.

### DISCUSSION THREAD

**Discussion by ELLIOTTCABLE** [ocamllabs/vscode-ocaml-platform#1091] [Ideas]
[These](https://github.com/ocaml/merlin/blob/183555e4d0b2fa7123f6141b1a2f32fd02d6795f/vim/merlin/doc/merlin.txt#L60-L7) were one of Merlin's best features, for working in OCaml code every day:

```vimhelp
:MerlinGrowEnclosing                                    *:MerlinGrowEnclosing*

When |:MerlinTypeOf| has been called, select the smallest expression
containing the previously highlighted expression.

Bound to <LocalLeader>n by default in normal mode.

:MerlinShrinkEnclosing                                *:MerlinShrinkEnclosing*

When |:MerlinGrowEnclosing| has been called, revert to the previously selected
expression. (i.e. the largest expression, centered around the position where
|:MerlinTypeOf| was called, which is contained in the currently highlighted
expression).

Bound to <LocalLeader>p by default in normal mode.
```

I see that there's `ocamllsp/wrappingAstNode` [in the upstream repo](https://github.com/ocaml/ocaml-lsp/blob/master/ocaml-lsp-server/docs/ocamllsp/wrappingAstNode-spec.md); could that perhaps be exposed in the UI as a pair of hotkeys?

**Comment by rgrinberg** [comment]
Sure, it could be exposed. You'll need to manage the selection inside the vscode extension though.

**Comment by ulugbekna** [comment]
There's also a built-in command "Expand Selection", which uses the AST based on LPS method `textDocument/selectionRange`
